File: archive-triage-runner/python/src/archive_triage/csv_writer.py
"""Write a stream of :class:`RowResult` values as CSV to an injected sink."""

from __future__ import annotations

import csv
from collections.abc import Iterable
from typing import TextIO

from .row_result import RowResult

HEADER: tuple[str, ...] = (
    "pedigree_id",
    "proband_name",
    "category",
    "refer_for_genetics",
    "triggers_matched_count",
    "error",
)


class CsvWriter:
    def __init__(self, sink: TextIO) -> None:
        self._writer = csv.writer(sink, lineterminator="\n")

    def write(self, rows: Iterable[RowResult]) -> None:
        self._writer.writerow(HEADER)
        for row in rows:
            self._writer.writerow(_row_values(row))


def _row_values(row: RowResult) -> tuple[str, str, str, str, str, str]:
    return (
        row.pedigree_id,
        row.proband_name,
        row.category,
        _format_bool(row.refer_for_genetics),
        str(row.triggers_matched_count),
        row.error,
    )


def _format_bool(value: bool | None) -> str:
    if value is None:
        return ""
    return "true" if value else "false"

File: archive-triage-runner/dotnet/src/ArchiveTriage/CsvWriter.cs
using System.Buffers;

namespace ArchiveTriage;

public sealed class CsvWriter
{
    public static readonly IReadOnlyList<string> Header = new[]
    {
        "pedigree_id",
        "proband_name",
        "category",
        "refer_for_genetics",
        "triggers_matched_count",
        "error",
    };

    private static readonly SearchValues<char> CharactersRequiringQuoting = SearchValues.Create(",\"\n\r");

    private readonly TextWriter sink;

    public CsvWriter(TextWriter sink)
    {
        this.sink = sink;
    }

    public async Task WriteAsync(IAsyncEnumerable<RowResult> rows)
    {
        await this.sink.WriteLineAsync(string.Join(',', Header.Select(EscapeField))).ConfigureAwait(false);
        await foreach (var row in rows.ConfigureAwait(false))
        {
            await this.sink.WriteLineAsync(FormatRow(row)).ConfigureAwait(false);
        }
    }

    private static string FormatRow(RowResult row) => string.Join(
        ',',
        new[]
        {
            EscapeField(row.PedigreeId),
            EscapeField(row.ProbandName),
            EscapeField(row.Category),
            EscapeField(FormatBool(row.ReferForGenetics)),
            EscapeField(row.TriggersMatchedCount.ToString(System.Globalization.CultureInfo.InvariantCulture)),
            EscapeField(row.Error),
        });

    private static string FormatBool(bool? value) => value switch
    {
        null => string.Empty,
        true => "true",
        false => "false",
    };

    private static string EscapeField(string value)
    {
        if (value.AsSpan().IndexOfAny(CharactersRequiringQuoting) < 0)
        {
            return value;
        }
        return "\"" + value.Replace("\"", "\"\"", StringComparison.Ordinal) + "\"";
    }
}
